### PR TITLE
Create records from data frames

### PR DESCRIPTION
**Pull request by lazappi**
When finished this PR should:

- [x] Add a `from_df()` method to (Artifact) registries to create artifacts from data frames
- [x] Modify printing for `Record` classes to avoid API calls (needed for temporary records)
- [x] Add temporary record classes for records that have been created but not yet saved
- [x] Add a "default instance" mechanism that limits one instance to creating records
- [x] A a `delete()` method to records

Other Todos:

- [x] Clean up and style changes
- [x] Document new functions
- [ ] Add tests
- [x] Check CI passes
- [ ] Add saving to vignette (?)
- [x] Update functionality documentation
- [x] Update architecture documentation
- [x] Update CHANGELOG

Removed:

- [x] ~Allow loading CSV and TSV artifacts~
- [x] ~Allow loading Parquet artifacts~

### CODE REVIEW

File: R/Registry.R [hunk 1]
```diff
@@ -138,6 +138,29 @@ Registry <- R6::R6Class( # nolint object_name_linter
...
+    from_df = function(dataframe, key = NULL, description = NULL, run = NULL) {
+      if (private$.registry_name != "artifact") {
+        cli::cli_abort(
+          "Creating records from data frames is only supported for the Artifact registry"
+        )
+      }
+
+      check_requires("Creating records from data frames", "reticulate")
+
+      py_lamin <- reticulate::import("lamindb")
```
Comment (rcannood) [inline R/Registry.R:150]: Should we also use `check_requires` for python packages?

```suggestion
      check_requires("Creating records from data frames", "reticulate")
      check_requires("Creating records from data frames", "lamindb", "python")

      py_lamin <- reticulate::import("lamindb")
```

File: R/Registry.R [hunk 2]
```diff
@@ -346,3 +369,75 @@ Registry <- R6::R6Class( # nolint object_name_linter
...
+    )
+    if (inherits(value, "lnschema_core.models.Record")) {
+      value <- create_record_from_python(value, instance)
+    }
+    value
+  }) |>
+    setNames(fields)
+
+  record_class <- registry$get_record_class()
+  temp_record_class <- create_temporary_record_class(record_class)
```
Comment (rcannood) [inline R/Registry.R:407]: Do we need to run `create_temporary_record_class` every time we create a record?

Or could it also be added to the registry (e.g. `registry$get_transient_record_class()`)

File: R/Registry.R (hunk 1 above)
Comment (rcannood) [reply R/Registry.R:150]: Or

```suggestion
      check_requires("Creating records from data frames", "reticulate")
      check_requires("Creating records from data frames", "lamindb", "reticulate")
      py_lamin <- reticulate::import("lamindb")
```

Comment (lazappi) [reply R/Registry.R:150]: Possibly. I was going to change how/when we check for Python **lamin** anyway.

File: R/Registry.R (hunk 2 above)
Comment (lazappi) [reply R/Registry.R:407]: Yes, if we are happy to add a bunch of extra classes that will rarely be used. Maybe we only add them to the default instance?

Comment (rcannood) [reply R/Registry.R:445]: are there that many classes which might need to be generated?

alternatively, `registry$get_record_class()` could have a few optional arguments to choose whether the record class is transient or not. This in turn could use lazy loading to ensure that we don't generate the same classes over and over again

**Review by rcannood — COMMENTED** [review]
One more question: could you also update the list of features to include the functionality that this PR adds?

File: R/Instance.R [hunk 3]
```diff
@@ -1,4 +1,4 @@
-create_instance <- function(instance_settings) {
+create_instance <- function(instance_settings, is_default = FALSE, py_lamin = NULL) {
   super <- NULL # satisfy linter
 
   api <- InstanceAPI$new(instance_settings = instance_settings)
```
[the remark this answers is not in the document]
Comment (lazappi) [reply R/Instance.R]: I moved it but I deliberately wanted to check when the instance is created so users knows ASAP if they can't do some things. We could change that if you want though.

File: R/Instance.R [hunk 4]
```diff
@@ -246,9 +258,23 @@ Instance <- R6::R6Class( # nolint object_name_linter
...
+    #' @field is_default (`logical(1)`)\cr
+    #' Whether this is the default instance.
+    is_default = function() {
+      private$.is_default
+    },
+    #' @field py_lamin (`python.builtin.module`)\cr
+    #' Python lamindb module.
+    py_lamin = function() {
+      private$.py_lamin
+    }
```
[the remark this answers is not in the document]
Comment (lazappi) [reply R/Instance.R]: Moved. I'm not sure it actually helps to store the imported module so we can could also load it as needed.

**Comment by lazappi** [comment]
> One more question: could you also update the list of features to include the functionality that this PR adds?

Added to the README. I already did the development vignette. Let me know if there is anywhere else.

**Pull request merged**